python/demo/demo_pml.py

```diff
@@ -0,0 +1,225 @@
+
+def curl_2d(a):
+    """Returns the curl of two 2D vectors as a 3D vector"""
+    ay_x = a[1].dx(0)
+    ax_y = a[0].dx(1)
+    return as_vector((0, 0, ay_x - ax_y))
+
+def pml_coordinates(x, alpha, k0, l_dom, l_pml):
+
+    # Define boolean functions for the PML regions
+    inside_pml_x = (ufl.sign(ufl.sign(x[0])*x[0] - l_dom/2) + 1)/2
+    inside_pml_y = (ufl.sign(ufl.sign(x[1])*x[1] - l_dom/2) + 1)/2
+
+    # Define the coordinate transformation for PML regions
+    x_pml = x[0] + 1j*alpha/k0*x[0]*(ufl.sign(x[0])*x[0] - l_dom/2)/(l_pml/2 - l_dom/2)**2*inside_pml_x
+    y_pml = x[1] + 1j*alpha/k0*x[1]*(ufl.sign(x[1])*x[1] - l_dom/2)/(l_pml/2 - l_dom/2)**2*inside_pml_y
+
+    return as_vector((x_pml, y_pml))
+
+um = 10**-6  # micron
+nm = um*10**-3  # nanometer
+epsilon_0 = 8.8541878128 * 10**-12
+mu_0 = 4 * np.pi * 10**-7
+
+# Radius of the wire and of the boundary of the domain
+radius_wire = 0.05 * um
+l_dom = 0.4 * um
+l_pml = 0.6 * um
+
+# The smaller the mesh_factor, the finer is the mesh
+mesh_factor = 1
+
+# Mesh size inside the wire
+in_wire_size = mesh_factor * 6 * nm
+
+# Mesh size at the boundary of the wire
+on_wire_size = mesh_factor * 3 * nm
+
+# Mesh size in the background
+bkg_size = mesh_factor * 20 * nm
+
+# Mesh size at the boundary
+pml_size = mesh_factor * 20 * nm
+
+# Tags for the subdomains
+au_tag = 1
+bkg_tag = 2
+pml_tag = 3
+# -
+
+model = generate_mesh_wire(
+    radius_wire, l_dom, l_pml, in_wire_size, on_wire_size, bkg_size,
+    pml_size, au_tag, bkg_tag, pml_tag)
+
+mesh, cell_tags, facet_tags = model_to_mesh(
+    model, MPI.COMM_WORLD, 0, gdim=2)
+
+gmsh.finalize()
+MPI.COMM_WORLD.barrier()
+# -
+
+wl0 = 0.4 * um  # Wavelength of the background field
+n_bkg = 1  # Background refractive index
+eps_bkg = n_bkg**2  # Background relative permittivity
+k0 = 2 * np.pi / wl0  # Wavevector of the background field
+deg = np.pi / 180
+theta = 0 * deg  # Angle of incidence of the background field
+
+degree = 3
+curl_el = ufl.FiniteElement("N1curl", mesh.ufl_cell(), degree)
+V = fem.FunctionSpace(mesh, curl_el)
+
+f = BackgroundElectricField(theta, n_bkg, k0)
+Eb = fem.Function(V)
+Eb.interpolate(f.eval)
+
+x = ufl.SpatialCoordinate(mesh)
+
+alpha = 1
+r_pml = pml_coordinates(x, alpha, k0, l_dom, l_pml)
+J_pml = grad(r_pml)
+
+# Transform the 2x2 Jacobian into a 3x3 matrix.
+J_pml = as_matrix(((J_pml[0, 0], 0          , 0),
+                   (0          , J_pml[1, 1], 0),
+                   (0          , 0          , 1)))
+
+A_pml = inv(J_pml)
+pml_matrix = det(J_pml) * A_pml * transpose(A_pml)
+eps_pml = eps_bkg * pml_matrix
+mu_pml = inv(pml_matrix)
+
+Es = ufl.TrialFunction(V)
+v = ufl.TestFunction(V)
+
+Es_3d = as_vector((Es[0], Es[1], 0))
+v_3d = as_vector((v[0], v[1], 0))
+
+dx = ufl.Measure("dx", mesh, subdomain_data=cell_tags)
+dAu = dx(au_tag)
+dBkg = dx(bkg_tag)
+dDom = dAu + dBkg
+dPml = dx(pml_tag)
+
+reps_au = -1.0782
+ieps_au = 5.8089
+eps_au = reps_au + ieps_au * 1j
```

Comment:
Define this as a single number, and use `np.real` and `np.imag` when you explicitly need the individual components.